File: concom.cpp
/*
ID: totheso1
LANG: C++14
TASK: concom
*/

#include <algorithm>
#include <fstream>
#include <iterator>
#include <limits>
#include <stack>
#include <vector>

using namespace std;

static ifstream fin("concom.in");
static ofstream fout("concom.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static constexpr int MAX_COMPANY = 100;
static vector<vector<int>> own(MAX_COMPANY, vector<int>(MAX_COMPANY, 0));
static vector<vector<bool>> control(MAX_COMPANY,
                                    vector<bool>(MAX_COMPANY, false));

static void initialize() {
  for (int i = 0; i < MAX_COMPANY; ++i) {
    control[i][i] = true;
  }
}

static void solve() {
  int line_count = fin_get<int>();
  while (line_count--) {
    const int i = fin_get<int>();
    const int j = fin_get<int>();
    const int percentage = fin_get<int>();
    own[i - 1][j - 1] = percentage;
    // own[fin_get<int>() - 1][fin_get<int>() - 1] = fin_get<int>();  // This is
    // wrong, because the evaluation order of function might not be what you
    // think.
  }

  for (int root = 0; root < MAX_COMPANY; ++root) {
    stack<int> stk({root});
    vector<vector<int>> total_percentage(MAX_COMPANY,
                                         vector<int>(MAX_COMPANY, 0));
    while (!stk.empty()) {
      const int curr = stk.top();
      stk.pop();
      for (int other = 0; other < MAX_COMPANY; ++other) {
        if (!control[root][other]) {
          total_percentage[root][other] += own[curr][other];
          if (total_percentage[root][other] > 50) {
            control[root][other] = true;
            stk.push(other);
          }
        }
      }
    }
  }

  int i = -1;
  for (const auto &row : control) {
    ++i;
    int j = -1;
    for (const bool is_controlling : row) {
      ++j;
      if (is_controlling && i != j) {
        fout << i + 1 << ' ' << j + 1 << endl;
      }
    }
  }
}

int main() {
  initialize();
  solve();
}

File: concom.py
"""
ID: totheso1
LANG: PYTHON3
TASK: concom
NOTE:
    I find this solution in USACO analysis is more complex
    and harder to prove and debug.
    It isn't necessarily faster than
    the "read all and find all" method below, too.
    So I'll stick with my method for now.
"""

from typing import Any

fin = open("concom.in", "r")
fout = open("concom.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
MAX_COMPANY = 100
own = [[0] * MAX_COMPANY for _ in range(MAX_COMPANY)]
control = [[False] * MAX_COMPANY for _ in range(MAX_COMPANY)]
for i in range(MAX_COMPANY):
    control[i][i] = True


def add_control(i: int, j: int) -> None:
    if control[i][j]:
        return
    control[i][j] = True
    for k in range(MAX_COMPANY):
        own[i][k] += own[j][k]
    for k in range(MAX_COMPANY):
        if control[k][i]:
            add_control(k, j)
    # I don't understand why this third loop
    # can't be placed before the second loop,
    # or combined in the first loop.
    for k in range(MAX_COMPANY):
        if own[i][k] > 50:
            add_control(i, k)


def add_own(i: int, j: int, percentage: int) -> None:
    for k in range(MAX_COMPANY):
        if control[k][i]:
            own[k][j] += percentage
            if own[k][j] > 50:
                add_control(k, j)


fin.readline()
for line in fin:
    i, j, percentage = map(int, line.split())
    # add_own(i - 1, j - 1, percentage)
    own[i - 1][j - 1] = percentage

for root in range(MAX_COMPANY):
    stack = [root]
    total_percentage = [[0] * MAX_COMPANY for _ in range(MAX_COMPANY)]
    while stack:
        curr = stack.pop()
        for other in range(MAX_COMPANY):
            if not control[root][other]:
                total_percentage[root][other] += own[curr][other]
                if total_percentage[root][other] > 50:
                    control[root][other] = True
                    stack.append(other)

for i, row in enumerate(control):
    for j, is_controlling in enumerate(row):
        if is_controlling and i != j:
            fprint(i + 1, j + 1)
# Code end
